debian/contrib/in.uucpd.8
.TH UUCPD 8 "Apr 20, 1999" "" "Linux Administrator's Manual"
.SH NAME
uucpd, in.uucpd \- run uucico from inetd
.SH SYNOPSIS
.B /usr/sbin/in.uucpd
.SH DESCRIPTION
.B Uucpd
is a program run from
.B inetd
when inetd detects a connect on the uucp port (usually
.I 540
). It uses pam to authenticate users. If access by pam would be granted,
and the first six letters of the basename of the shell
match the word
.I uucico
then an utmp and wtmp entry are written and the users shell is invoked.
The uucpd process stays around until
.B uucico
exits, then updates the utmp and wtmp files once more.

.SH OPTIONS
None.
.SH NOTES
Uucico also has the possibility to check for loginname/password itself
directly, instead of needing a frontend uucpd. However that does not
work if the system has shadow passwords, or any other way of authentication
which requires root privileges. Also, no utmp/wtmp records are written.
.PP
If you want to define UUCP accounts outside the password file (eg
through
.I /etc/uucp/passwd
) you need to call
.B uucico -l
from
.I /etc/inetd.conf
instead of
.B in.uucpd
\.

.SH AUTHOR
Miquel van Smoorenburg (miquels@cistron.nl)
.SH "SEE ALSO"
.BR inetd (8),
.BR inetd.conf (5),
.BR uucico (8)


debian/changelog
uucp (1.07-30) unstable; urgency=medium

  * fix issues with gcc14 (Closes: #1075610)

 -- Thorsten Alteholz <debian@alteholz.de>  Sun, 28 Jul 2024 13:00:00 +0200

uucp (1.07-29) unstable; urgency=medium

  * take care of implicit function declaration (Closes: #1066434)
    (this even affects the configure script)

 -- Thorsten Alteholz <debian@alteholz.de>  Fri, 12 Apr 2024 22:00:00 +0200

uucp (1.07-28) unstable; urgency=medium

  * Fix FTBFS when dh_installsystemd moves units to /usr
    Thanks to Helmut Grohne for the patch
    (Closes: #1052716)
  * debian/control: bump standard to 4.6.2 (no changes)
  * debian/patches: no need to forward them to upstream

 -- Thorsten Alteholz <debian@alteholz.de>  Tue, 03 Oct 2023 18:00:00 +0200

uucp (1.07-27) unstable; urgency=medium

  * add improved-pipe.patch to be able to replace the placeholder \Z
    by the system name of the called system
    thanks to Rafael Diniz for the patch

 -- Thorsten Alteholz <debian@alteholz.de>  Sat, 17 Aug 2019 22:00:00 +0200

uucp (1.07-26) unstable; urgency=medium

  * debian/control: add Salsa VCS URLs
  * debian/control: no longer Depend: on inetd but only Recommends:
                    (Closes: #589535)
  * debian/control: no longer Depend: on cu but only Recommends:
  * debian/uucp.postinst: only call update-inetd when available
                    (Closes: #919277)
  * debian/uucp.postinst: remove code for old version no longer
                          available in any release
  * debian/uucp.postrm: avoid paths
  * add service file for uucp, don't automatically use it
    (thanks to Jörg Sommer for the service file)
  * add patch to wait for correct parent (Closes: #922438)
    (thanks to Jörg Sommer for the patch)

 -- Thorsten Alteholz <debian@alteholz.de>  Sat, 02 Mar 2019 16:00:00 +0100

uucp (1.07-25) sid; urgency=medium

  * debian/control: use dh11
  * debian/control: bump standard to 4.3.0 (no changes)
  * debian/control: no longer depend on exim
  * debian/control: add ORed dependency on cron-daemon (Closes: #877059)
                    Thanks to Jörg Sommer for the hint.
  * debian/rules: add hardening flags
  * Fix FTCBFS: (Closes: #879580)
    + Thanks to Helmut Grohne for the patch!
    + Let dh_auto_build pass cross compilers to make in contrib.
    + Use a triplet-prefixed CC in d/rules.
  * debian/patches: improve spelling.patch
  * debian/patches: improve manpage.patch (Closes: #902671)
                    Thanks to Patrik Schindler for the hint.


 -- Thorsten Alteholz <debian@alteholz.de>  Thu, 10 Jan 2019 19:00:00 +0100

uucp (1.07-24) unstable; urgency=medium

  * debian/control: bump standard to 4.1.0 (no changes)
  * debian/control: use dh10
  * debian/control: set priority to optional
  * add option-to-not-force-hw-control.patch (Closes: #336996)
    introduce option to disable hw control on serial line
    thanks to Michal Suchanek for the patch
  * avoid unnecessary sleeps (Closes: #760556, #675101)
    thanke to Frank Heckenbach for the patch
  * no longer depends on install-info (Closes: #724779)
    (see 1.07-22 and #598904 for more info)
  * keeping /etc/cron.d/uucp (Closes: #473203)
  * Recommends: instead of Depends: on logrotate (Closes: #422969)
  * fix spelling error in debian/contrib/in.uucpd.8

 -- Thorsten Alteholz <debian@alteholz.de>  Wed, 27 Sep 2017 19:00:00 +0200

uucp (1.07-23) unstable; urgency=medium

  * upload to unstable now
  * confirm NMU
  * patch from NMU also closes: #759285
    with patch from Marco d'Itri

 -- Thorsten Alteholz <debian@alteholz.de>  Thu, 08 Oct 2015 19:00:00 +0200

uucp (1.07-20.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Pass "-s /bin/sh" to "su uucp" to cope with the change of uucp's shell
    in base-passwd 3.5.30 (closes: #734723).

 -- Colin Watson <cjwatson@debian.org>  Fri, 17 Jan 2014 12:44:41 +0000

uucp (1.07-22) experimental; urgency=low

  * debian/README.Debian: correct path to info files (Closes: #453957)
  * debian/README.exim4: update pointer to exim4 config
  * debian/control: fix typo (Closes: #602629)
  * debian/control: add 'serial' to description of cu (Closes: #358861)
  * debian/rules: call to deprecated install-info comes from debhelper
                  (thanks Guillem Jover for checking)
                  (Closes: #598904)

 -- Thorsten Alteholz <debian@alteholz.de>  Sat, 27 Jul 2013 12:00:00 +0200

uucp (1.07-21) experimental; urgency=low

  * New maintainer (Closes: #699118)
  * upload to experimental
  * debian/control: switch to dh 9
  * debian/control: switch to standard 3.9.4
  * debian/control: add homepage field
  * debian/control: no need to Conflict: or Replace: with uucp v1.06 anymore
  * debian/control: no need to Conflict: with tua 4.3-3 anymore
  * debian/control: add ${misc:Depends} to binary uucp package
  * debian/control: add bsd-mailx as real package to mailx dependency
  * debian/source/format: use 3.0 (quilt)
  * debian/copyright: change to machine readable format
  * debian/patches: add patch to fix manpage issues and spelling error
  * debian/watch: file added

 -- Thorsten Alteholz <debian@alteholz.de>  Sun, 23 Jun 2013 11:00:00 +0200

uucp (1.07-20) unstable; urgency=low

  * New maintainer                  Closes: #517906
    Thanks to Peter Palfrader and Christian Perrier for their
    prior work on these packages.
  * Update pointer to exim4 doc     Closes: #449172

 -- Richard A Nelson (Rick) <cowboy@debian.org>  Mon, 02 Mar 2009 22:28:00 -0000

uucp (1.07-19.3) unstable; urgency=low

  * Non-maintainer upload to fix an error in former NMU
  * Drop useless debian/po directory. Therefore, translations
    are now longer used. Closes: #482133
    With apologies to translators...

 -- Christian Perrier <bubulle@debian.org>  Sat, 02 Aug 2008 22:12:33 +0200

uucp (1.07-19.2) unstable; urgency=low

  * Non-maintainer upload.
  * Drop debconf note dealing with an obsolete transition. Therefore
    drop the .config script and dependency on debconf.

debian/contrib/in.uucpd.c
/*
 * uucpd.c	Designed to be run from inetd. Checks loginname/password,
 *		also checks if this _really_ is a uucp login.
 *		Does utmp/wtmp accounting.
 *
 * Version:	@(#)uucpd.c  1.00  19-May-1998  miquels@cistron.nl
 * Version:	@(#)uucpd.c  1.10  28-Jul-2001  alex@king.net.nz
 *
 */
#include <sys/types.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
#include <sys/wait.h>
#include <syslog.h>
#include <utmp.h>
#include <time.h>

#include <pwd.h>
#include <grp.h>

#include <security/pam_appl.h>
#include <security/pam_misc.h>

#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>

int myconv(int n,const struct pam_message **msg, struct pam_response **resp, void *ap)
{
	int k,rv;
	char *cp;

	rv=misc_conv(n,msg,resp,ap);
	if (rv != PAM_SUCCESS)
		return rv;
	/* 
	 * The supplied misc_conv function doesn't seem to strip the
	 * trailing return character so I do it here...
	 */
	for (k=0;k<n;k++)
		for (cp=((*resp)->resp);*cp!=0;cp++)
			if (*cp=='\r') *cp=0;
	return PAM_SUCCESS;
}

static struct pam_conv conv = {
	myconv,
	NULL
};

void chop(char *s)
{
	char *p;

	for(p = s; *p; p++)
		if (*p == '\r' || *p == '\n') {
			*p = 0;
			break;
		}
}


/*
 *	Read one line from the input.
 */
int getstr(int fd, char *buf, int len)
{
	int i, n, cnt;
	int nlseen = 0;

	cnt = 0;

	while(!nlseen) {
		if ((n = read(fd, buf + cnt, len - cnt)) <= 0) {
			if (n == -1 && errno == EINTR)
				continue;
			break;
		}
		cnt += n;
		for(i = 0; i < cnt; i++) {
			if (buf[i] == '\r' || buf[i] == '\n') {
				nlseen = 1;
				break;
			}
		}
		if (cnt >= len) break;
	}

	return cnt > 0 ? cnt : n;
}


int main(void)
{
	char login[UT_NAMESIZE];
	char host[UT_HOSTSIZE];
	struct passwd *pwd;
	char *s;
	struct sockaddr_in sin;
	int sinlen;
	struct hostent *h;
	struct utmp ut;
	pid_t pid;
	int st;
	int rv;
	pam_handle_t *pamh=NULL;

	/*
	 *	Make sure we have fds 0, 1 and 2.
	 */
	close(1);
	close(2);
	dup(0);
	dup(0);
	openlog("uucpd", LOG_PID, LOG_UUCP);

	/*
	 *	Get the remote host address.
	 */
	sinlen = sizeof(sin);
	if (getpeername(0, (struct sockaddr *)&sin, &sinlen) < 0) {
		syslog(LOG_ERR, "getpeername: %m");
		exit(1);
	}
	if ((h = gethostbyaddr((char *)&sin.sin_addr,
	    sizeof(sin.sin_addr), AF_INET)) != NULL) {
		strncpy(host, h->h_name, sizeof(host));
		host[sizeof(host) - 1] = 0;
	} else
		strcpy(host, inet_ntoa(sin.sin_addr));

	login[0] = 0;

	alarm(60);

	while(1) {
		printf("login: ");
		fflush(stdout);
		if (getstr(0, login, sizeof(login)) < 0)
			exit(1);
		login[sizeof(login) - 1] = 0;
		chop(login);
		if (login[0] != 0) break;
	}

	rv=pam_start("uucp",login,&conv,&pamh);
	if (rv==PAM_SUCCESS)
		rv=pam_authenticate(pamh, 0);
	if (rv==PAM_SUCCESS)
		rv=pam_set_item(pamh,PAM_RHOST,host);
	if (rv==PAM_SUCCESS) 
		rv=pam_acct_mgmt(pamh,0);
	if (rv!=PAM_SUCCESS) {
		syslog(LOG_AUTHPRIV|LOG_NOTICE,"%s",pam_strerror(pamh,rv));
		goto incorrect;
	}
	if ((pwd = getpwnam(login)) == NULL) {
incorrect:
		syslog(LOG_AUTHPRIV|LOG_NOTICE,
			"invalid password for `%s' on `TCP' from `%s'",
			login, host);
		printf("Login incorrect.\r\n");
		exit(1);
	}

	/*
	 *	See if we have a valid shell: the basename must
	 *	match uucico*
	 *
	 *	Note: arguably this check should be done by PAM, but
	 *	an appropriate PAM module doesn't exist to my knowledge.
	 */
	if ((s = strrchr(pwd->pw_shell, '/')) == NULL)
		s = pwd->pw_shell;
	else
		s++;
	if (strncmp(s, "uucico", 6) != 0) {
		syslog(LOG_AUTHPRIV|LOG_NOTICE,
			"invalid shell for `%s' on `TCP' from `%s'",
			login, host);
		printf("Invalid shell (not uucico)\r\n");
		exit(1);
	}

	/*
	 *	Allright we can assume the rest will go OK.
	 */
	memset(&ut, 0, sizeof(ut));
	ut.ut_type = LOGIN_PROCESS;
	strncpy(ut.ut_user, login, sizeof(ut.ut_user));
	strncpy(ut.ut_host, host, sizeof(ut.ut_host));
	ut.ut_addr = sin.sin_addr.s_addr;
	ut.ut_time = time(NULL);

	/*
	 *	Fork, let the parent wait to write the utmp file.
	 */
	signal(SIGHUP, SIG_IGN);
	if ((pid = fork()) != 0) {

		if (pid < 0) {
			/* FIXME: SCREAM (oh well..) */
			return 1;
		}
		pam_open_session(pamh, 0);

		ut.ut_pid = pid;
		sprintf(ut.ut_line, "uucp%d", ut.ut_pid);
		snprintf(ut.ut_id, sizeof(ut.ut_id), "uu%02x", ut.ut_pid&0xff);
		pututline(&ut);
		endutent();
		updwtmp(WTMP_FILE, &ut);

		while(wait(&st) != pid);
			;

		ut.ut_type = DEAD_PROCESS;
		ut.ut_time = time(NULL);
		ut.ut_user[0] = 0;
		pututline(&ut);
		endutent();
		updwtmp(WTMP_FILE, &ut);
		pam_close_session(pamh, 0);
		pam_end(pamh,PAM_SUCCESS);

		return 0;
	}

	/*
	 *	This is the child - exec uucico.
	 */
	signal(SIGHUP, SIG_DFL);
	if (chdir(pwd->pw_dir) != 0) {
		perror(pwd->pw_dir);
		exit(1);
	}
	if (setgid(pwd->pw_gid) != 0) {
		perror("setgid");
		exit(1);
	}
	if (initgroups(pwd->pw_name, pwd->pw_gid) < 0) {
		perror("initgroups");
		exit(1);
	}
	if (setuid(pwd->pw_uid) != 0) {
		perror("setuid");
		exit(1);
	}
	setenv("LOGNAME", pwd->pw_name, 1);
	setenv("USER", pwd->pw_name, 1);
	setenv("SHELL", pwd->pw_shell, 1);
	setenv("TERM", "dumb", 1);


	execl(pwd->pw_shell, pwd->pw_shell, NULL);

	perror(pwd->pw_shell);
	return 1;
}

